Code-First Federation 2 support

Mithras opened the pull request:
Subj.

Some details on what's in the PR:

1. Added [.devcontainer](https://github.com/Mithras/graphql-dotnet/tree/master/.devcontainer) because I don't want to install 9000 SDKs and packages on my host. OmniSharp (Intellisense) is half broken though because it works only with NET 6. This is not related to Federation and I can remove it if you don't want it. Or I can try to fix OmniSharp if you are interested in developing in containers.
2. Separate [GraphQL.Federation.Tests](https://github.com/Mithras/graphql-dotnet/tree/master/src/GraphQL.Federation.Tests) project with a couple INTEGRATION tests (run in 75ms though). I'm not sure you want integration tests and you might prefer 9000 unit tests instead so I made it separate so that it's easy to get rid of. I don't mind adding unit tests, just didn't want to waste time as I'm not sure this PR will ever get reviewed/merged. The tests are pretty simple and self explanatory, see [FederationFieldTests](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL.Federation.Tests/FederationFieldTests.cs).
3. A bunch of NEW stuff under [GraphQL.Federation](https://github.com/Mithras/graphql-dotnet/tree/master/src/GraphQL/Federation) namespace.
  3.1. `Attributes.*` - All Federation 2 [directives](https://www.apollographql.com/docs/federation/federated-types/federated-directives/) as attributes.
  3.2 [Extensions.FederationFieldExtensions](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationFieldExtensions.cs) - All Federation 2 [directives](https://www.apollographql.com/docs/federation/federated-types/federated-directives/) as `FieldType`/`FieldBuilder` extensions.
  3.3 [Extensions.FederationGraphTypeExtensions](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationGraphTypeExtensions.cs) - All Federation 2 [directives](https://www.apollographql.com/docs/federation/federated-types/federated-directives/) as `GraphType` extensions. Also, `.AddServices()` and `.AddEntities()` to add Federation fields explicitly/manually (e.g. when you want to specify `.AuthorizeWith()` on them).
  3.4 [Extensions.FederationGraphQLBuilderExtensions](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationGraphQLBuilderExtensions.cs) - `AddFederation()` extension. See [src/GraphQL.Federation.Tests/ServiceProviderFixture.cs](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL.Federation.Tests/ServiceProviderFixture.cs#L19).
  3.5 `Schema.*` - A few output types to support Federation.
  3.6. A bunch of `internal` classes.

There are a few questionable hacks:

1. [EntityType](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Schema/EntityType.cs) adds pointless [NeverType](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Schema/NeverType.cs) because that's the only way I can pass validation that triggers BEFORE I have a chance to add any Federated entities to the `EntityType` which I do in [FederationEntitiesSchemaNodeVisitor](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationEntitiesSchemaNodeVisitor.cs).
2. [FederationQuerySchemaNodeVisitor](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationQuerySchemaNodeVisitor.cs) adds `_service` and `_entities` in a bit of a weird way (injects GraphType instances in ctor) because it doesn't want to work otherwise. Probably something about order of when schema initializes and visitors are called.
3. [FederationQuerySchemaNodeVisitor.ResolveEntities()](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationQuerySchemaNodeVisitor.cs#L40) doesn't check authorization at an entity level. Only at a field level.
4. **[Ignore]** _I also copy-pasted [GraphTypeToClrType()](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationHelper.cs#L136) from our code which I feel like should be provided by `graphql-dotnet` by default but I'm not aware of any other ways of mapping Graph <-> CLR types. It's used by [FederationQuerySchemaNodeVisitor.ResolveEntities()](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationQuerySchemaNodeVisitor.cs#L40)._

Hope this won't get stuck for years like https://github.com/graphql-dotnet/graphql-dotnet/pull/1669 did.

Shane32 commented:
At first glance I like the look of it!

joemcbride commented:
Just popping in for a moment. This does look pretty interesting.

A reminder that there are already some federation types that look almost exactly the same as these used with schema-first. Probably want to figure out some reuse there.

https://github.com/graphql-dotnet/graphql-dotnet/tree/master/src/GraphQL/Utilities/Federation

Mithras commented:
> A reminder that there are already some federation types that look almost exactly the same as these used with schema-first. Probably want to figure out some reuse there.

I think I re-used what I can (only `AnyScalarGraphType` really). The rest has breaking changes (I'm not sure how existing Federation is used if at all). I can merge them if we are ok with breaking changes:
- Existing `FederatedSchemaPrinter ` doesn't support new Federation 2 directives and doesn't add `@link` directive to schema. You can say that existing one is for Federation 1 and the new one is for Federation 2.
- The main difference between existing [IFederatedResolver](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/Federation/IFederatedResolver.cs) and my [IFederationResolver](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/IFederationResolver.cs) is that the later supports all 3 sync, async and DataLoader scenarios while the existing one ONLY supports async. See 3 different overloads [here](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Extensions/FederationGraphTypeExtensions.cs#L76)
- Existing [ServiceGraphType](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/Federation/ServiceGraphType.cs) uses Federation 1 printer, while my [ServiceGraphType](https://github.com/Mithras/graphql-dotnet/blob/master/src/GraphQL/Federation/Schema/ServiceGraphType.cs) uses Federation 2 printer.
- The rest under [Utilities/Federation](https://github.com/graphql-dotnet/graphql-dotnet/tree/master/src/GraphQL/Utilities/Federation) are schema-first specific and I don't need/have them.

Also, I'm not sure what's the story with schema-first approach in general. Is it something graphql-dotnet actively supports?

joemcbride commented:
My point was - I think it would be fairly awkward to have two different implementations of federation in the same project. If we're upgrading federation support to add GraphType First support, it should be upgraded everywhere and not work differently when using Schema First vs. GraphType First.

> Also, I'm not sure what's the story with schema-first approach in general. Is it something graphql-dotnet actively supports?

I certainly hope so since I actively use it. 😉

Schema First at it's core uses the same GraphQL Framework. It is just a light schema builder on top of the core engine. From that regard, it is basically the same as the `AutoRegisteringGraphType`. That is where the `GraphQLAttribute` was first conceived and added, which is heavily used by this PR.

Mithras commented:
@joemcbride Let me first look into supporting schema-first in my implementation and then we can decide on either merging or deprecating the old one completely.

Mithras commented:
@joemcbride do you know why I see this changes when running tests locally but in CI they are not present?
`/workspaces/graphql-dotnet/src/GraphQL.ApiTests/netstandard20+netstandard21/GraphQL.approved.txt`
![image](https://user-images.githubusercontent.com/2098029/169373309-0c19e636-e7ee-4d96-ab26-97c0cb2f8c61.png)

Mithras commented:
Oh, ok. I rolled back these lines for now. I'll re-run tests after rebasing.

Shane32 commented:
> I also see these changes which also seems to be VS Code version related

I can't help there; I never use VS Code.  As far as I know, @sungam3r does not either; but perhaps @joemcbride does.

Shane32 commented on src/GraphQL.Federation.Tests/FederationFieldTests.cs:
```diff
@@ -0,0 +1,77 @@
+}";
+        }).ConfigureAwait(false);
+
+        json.ShouldBe("{\"data\":{\"_entities\":[{\"name\":\"111\",\"externalTest\":{\"id\":4},\"externalResolvableTest\":{\"id\":7,\"external\":\"external-7\",\"extended\":\"external-7\"}},{\"name\":\"333\",\"externalTest\":{\"id\":6},\"externalResolvableTest\":{\"id\":9,\"external\":\"external-9\",\"extended\":\"external-9\"}}]}}");
```
Maybe `ShouldBeCrossPlatJson` here?

Mithras replied:
yes, let me try that

Shane32 commented on src/GraphQL.Federation.Tests/FederationFieldTests.cs:
```diff
@@ -0,0 +1,77 @@
+            .GetProperty("sdl")
+            .GetString();
+
+        sdl.ShouldBe("schema {\n  query: TestQuery\n}\n\ntype DirectivesTestDto @key(fields: \"id\") @shareable @inaccessible {\n  id: Int!\n  shareable: String @shareable\n  inaccessible: String @inaccessible\n  override: String @override(from: \"OtherSubgraph\")\n  external: String @external\n  provides: String @provides(fields: \"foo bar\")\n  requires: String @requires(fields: \"foo bar\")\n} ...
```
There's some way to tell Shouldly to ignore line ending differences... can't remember how at the moment.  Probably another argument.

Mithras replied:
I'll try to use `ShouldBeCrossPlat` from https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL.Tests/AssertionExtensions.cs

Mithras commented:
I've added schema-first Federation support. Please review and let me know if it's ok to remove existing Federation classes.

Mithras commented:
bump

sungam3r commented:
Hi. Now I am participating in the review of the server project - WebSockets/subscriptions redesign. After that I could review this one. I began my own federation rework some time ago too, but abandoned.

Mithras commented:
Added `SchemaPrinterOptions` to `.AddFederation()` extension so that it's possible to expose `@deprecated` directives and/or descriptions from subgraphs.

Mithras commented:
Fixed KeyAttribute to allow multiple keys.

Mithras commented:
bump

sungam3r commented:
Yeah, I see all your commits in my email notifications. Sorry, still busy with reviewing other stuff in preparation for v7.

sungam3r commented:
This PR is the only one in my email and I see it on top of the list of letters every time when I watch email. Still has no enough time to deal with all of that.

Mithras commented:
Alright, sorry for spamming. This PR is pretty much what we use today in an internal company package and I just really want to merge this into graphql-dotnet so that I don't need to maintain the same code in two places. Plus it's less likely to break in v7 or whatever is coming when it's part on graphql-dotnet

sungam3r commented:
@Mithras FYI. I have one PR with full rewrite of schema printing (not published yet). I started it for a long time ago and the only thing that stopped me was changes in printing of federated schemas. Our PRs intersect in this part. I did not set my goal of rewriting federalization.

sungam3r commented:
I'll return to this PR after v7 release.

hwillson commented:
👋 @sungam3r - is there anything we can do to help expedite this PR? This is really awesome work - thanks for putting this together @Mithras!

sungam3r commented:
I'll review it some time after v7 release. It definetely should be released in v8.

wgebczyk commented on src/GraphQL/Federation/Extensions/FederationHelper.cs:
```diff
@@ -0,0 +1,128 @@
+                                            .Cast<FederationDirectiveEnum>()
+                                            .Select(x => new GraphQLStringValue(FederationDirectiveEnumMap[x]))
+                                            .Cast<GraphQLValue>()
+                                            .ToList()
```
This could be written without GetFlags helper as:
```
Values =
    Enum.GetValues(typeof(FederationDirectiveEnum))
    .Cast<FederationDirectiveEnum>()
    .Where(x => x.HasFlag(import))
    .Select(x => new GraphQLStringValue(FederationDirectiveEnumMap[x]))
    .Cast<GraphQLValue>()
    .ToList()
```

Mithras commented:
Rebased on latest master

daniiltkach-ms commented:
Hey guys! First of all, i wanna say thank you for everything what you are doing. this stuff is so great!
Also i would like to know when you gonna merge this pr ? :D

Shane32 commented:
> I'll review it some time after v7 release. It definetely should be released in v8.

Ping @sungam3r - If you have time available now, I suggest focusing it on this PR.

sungam3r commented:
Yeah, I remember, really. I need time to sort ~500 emails of "high/moderate priority" after my break and after that I will return to postponed things like that. This is a big one I must admit. What I can say with 100% confidence is that since now (the second half of October as I said sometime earlier) and till spring I indeed will have time to review/post PRs. Just need a bit time to tune in to work and remember abandoned business.

sungam3r commented:
260 letters and a couple of PRs to review left. I think I'll come back to this one next week.

sungam3r commented:
For information - 130 left, I'm still fighting with email and code review. Should start to work on that issue at the weekend.

sungam3r commented on src/GraphQL/Federation/Extensions/FederationGraphQLBuilderExtensions.cs:
```diff
@@ -0,0 +1,48 @@
+namespace GraphQL.Federation.Extensions;
+
+/// <summary>
+/// Federation extensions for IGraphQLBuilder
```
Please, use `<see cref` constructs when referencing types/methods.

Shane32 replied:
> ends tests with dot

Meaning: End xml comments with periods, such as the following:

```suggestion
/// Federation extensions for <see cref="IGraphQLBuilder" />.
```

sungam3r commented on src/GraphQL.ApiTests/net6/GraphQL.approved.txt:
```diff
@@ -1168,7 +1168,6 @@ namespace GraphQL.Execution
     public class ExecutionStrategyRegistration : System.IEquatable<GraphQL.Execution.ExecutionStrategyRegistration>
     {
         public ExecutionStrategyRegistration(GraphQL.Execution.IExecutionStrategy Strategy, GraphQLParser.AST.OperationType Operation) { }
-        protected virtual System.Type EqualityContract { get; }
```
Rebase onto master to resolve this issue.

sungam3r commented:
Well, in recent days I read and sorted all the accumulated mail in this repository and I only have left this one issue to work on. I did not have time to start working on this on the weekend, but I will definitely start tomorrow. I had a very-very brief look into your code. My suggestions so far:

1. Rebase onto `master` to "fresh" your code.
2. Switch to NET7 into new projects and dockerfiles.
3. Rename `Ql` into `QL` in all names like `BaseSchemaFirstGraphQlTest`.
4. Do not use `Exception` base class in all `throw` statements.

This is a rather large PR and most likely it will be better to break it into parts. For example:
1. New attributes for federation / directives / core primitives.
2. Implementation.
3. Tests.
4. Documentation. Documentation is no less important than the code itself.

I also see more directives (7) than those that I once read (4). I will have to dive into the theoretical part.

And more thoughts about responsibility zones and future development. 
1) I think this PR will inevitably make backward incompatible changes so maybe retarget to `develop` for v8? 
2) What if move all federation stuff into separate project? Not everyone needs federation. A separate package will allow to localize APIs and make it more focused.

sungam3r commented:
Another reason I think about moving federation into separate package is that federation is essentially a third-party (by Apollo) extension on top of the core GraphQL spec. I would not have a third-party bits into main project. Generally speaking, the same thing I think about the `UseApolloTracing` method and classes it uses. For me it would be good to bring it into a separate package as well (the same as federation or another one), if the entire code for its implementation would not be ~100 lines of code.

> I will have to dive into the theoretical part.

@Mithras What resources could you recommend to read about federation?

dariuszkuc commented:
Hello 👋 
re: fed docs -> for Apollo Federation overview I'd recommend to take a look at https://www.apollographql.com/docs/federation/ and https://www.apollographql.com/docs/federation/subgraph-spec for the subgraph changes.

Shane32 commented:
> I think this PR will inevitably make backward incompatible changes so maybe retarget to `develop` for v8?

At this time, master and develop point to the same commit.  We can work off of master for now, and then switch to develop if there are breaking changes.

> What if move all federation stuff into separate project? Not everyone needs federation. A separate package will allow to localize APIs and make it more focused.

This makes sense to me -- if there is a lot of code necessary to properly implement federation, and if it can be reasonably implemented as a layer over the base GraphQL.NET codebase.  If not, I'd just keep it in the same package.

Also, if the changes are all in a new package, perhaps there is no need to target v8.

hwillson commented:
Moving the Federation stuff into a separate project is definitely a viable (and sometimes preferred) approach. We follow a similar approach ourselves with [Federation JVM](https://github.com/apollographql/federation-jvm/) (provides Federation support for `graphql-java`) and [`@apollo/subgraph`](https://github.com/apollographql/federation/tree/main/subgraph-js) (provides Federation support for `graphql-js` based projects).

Mithras commented:
- Rebased on master
- Upgraded dev container to .NET 7.0
- Re-run `dotnet build/test/format` (two unrelated tests fail for me even in `graphql-dotnet/master`. See [Bug2194.cs](https://github.com/graphql-dotnet/graphql-dotnet/pull/3144/files#diff-82e09709755198a23e74b4958263f7e9775ce58ceba6388b648e61aa18042e03) and [SchemaPrinterTests.cs](https://github.com/graphql-dotnet/graphql-dotnet/pull/3144/files#diff-f4b4db45fb999064e4b918128847dd2d3965a7b3de4f1676c32a1f4fb23a7162))
- Fixed some xml comments
- Renamed all "Ql" to "QL"
- Got rid of `throw new Exception(...)`

> This is a rather large PR and most likely it will be better to break it into parts

I don't mind splitting but this is minimal working changes. I.e. there is no value in having attributes that don't do anything. And splitting implementation from tests also sound a bit strange to me. It's up to you though, I can do either.

> I think this PR will inevitably make backward incompatible changes so maybe retarget to develop for v8?

Personally I don't think it matters because current implementation is not usable in production and I really doubt anybody is using it. E.g. no `DataLoader` support which makes it unusable outside of "hello world" federation. 

> What if move all federation stuff into separate project? Not everyone needs federation. A separate package will allow to localize APIs and make it more focused.

Just to clarify, separate project in the repo or completely separate like ASP.NET integration from @Shane32 ? Separate project is fine. Separate repo is what I'm trying to avoid here as having it as part of the main repo ensures things are not broken in new releases (e.g. there is no printer changes that make Federation impossible without copy-pasting some classes). I need to check but I think I'm also using a couple internal methods.

> What resources could you recommend to read about federation?

links from @dariuszkuc are good.

Also, I'm traveling right now and probably won't have time for bigger refactorings like moving to a separate project and/or splitting until next week.

Shane32 commented on src/GraphQL/Federation/Extensions/FederationGraphTypeExtensions.cs:
```diff
@@ -0,0 +1,110 @@
+        astMetadata!.Directives!.Items.Add(directive);
+    }
+
+    /// <summary>
+    /// Specifies reference resolver for the type.
+    /// </summary>
+    public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, TSourceType> resolver) =>
+        graphType.Metadata[RESOLVER_METADATA] = new FederationResolver<TSourceType>(resolver);
+
+    /// <summary>
+    /// Specifies reference resolver for the type.
+    /// </summary>
+    public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, Task<TSourceType>> resolver) =>
+        graphType.Metadata[RESOLVER_METADATA] = new AsyncFederationResolver<TSourceType>(resolver);
+
+    /// <summary>
+    /// Specifies reference resolver for the type.
+    /// </summary>
+    public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, IDataLoaderResult<TSourceType>> resolver) =>
+        graphType.Metadata[RESOLVER_METADATA] = new DataLoaderFederationResolver<TSourceType>(resolver);
```
See here, where we include extension methods for up to 3 levels of data loaders, for both synchronous and asynchronous resolvers:
- https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL.DataLoader/Extensions/DataLoaderExtensions.cs#L173

sekizi replied:
It's been too long but as I remember I had one class initially and then switched to this becausew it was making something else easier. I'll try again.

Mithras replied:
I simplified it to some extend. I got rid of multiple resolver classes and kept only one `FuncFederationResolver<T>`. Made both `IFederationResolver` and `FuncFederationResolver<T>` and added extensions to pass any `IFederationResolver`. What I wasn't able to do is making it `ValueTask<object>` instead of just `object` (see above).

sekizi commented on src/GraphQL/Federation/IFederationResolver.cs:
```diff
@@ -0,0 +1,7 @@
+internal interface IFederationResolver
+{
+    Type SourceType { get; }
+    object Resolve(IResolveFieldContext context, object source);
```
Sure but 99% of real world cases it'll return DataLoader.

Mithras replied:
`ValueTask<object>` doesn't work out of the box (something is failing when `IDataLoaderResult` is wrapped in `ValueTask`). It's used like `new FieldType { ..., Resolver = new FuncFieldResolver<object>(items) }` where `items` are populated from the `Resolve()` above. It works fine when some items are `IDataLoaderResult` but throws some validation errors when items are `ValueTask<IDataLoaderResult>`

sungam3r commented:
>  OmniSharp (Intellisense) is half broken though because it works only with NET 6. 

How does it feel on NET7 ?

> Or I can try to fix OmniSharp if you are interested in developing in containers.

I must admit I have no experience in containers at all but I always wanted to learn.

> Separate [GraphQL.Federation.Tests](https://github.com/Mithras/graphql-dotnet/tree/master/src/GraphQL.Federation.Tests) project with a couple INTEGRATION tests (run in 75ms though).

Fine with me.

> My point was - I think it would be fairly awkward to have two different implementations of federation in the same project. If we're upgrading federation support to add GraphType First support, it should be upgraded everywhere and not work differently when using Schema First vs. GraphType First.

Agree with @joemcbride . I will prefer to accept backward incompatible changes than to have two different implementations on support.

Regarding VS Code configuration files. Yes, I don't use VS Code and I'm fine to change these files in any way you see fit. Or simply remove them from this repo and add into gitignore.

Regarding unit tests with spaghetti string literals. I recommend you to switch to raw string literals that are available since C#11.

> I've added schema-first Federation 2 support. Can you please review and let me know if it's ok to remove existing Federation classes?

I must admit that this is my personal opinion and I do not use federation and I still know little about it. If Apollo recommends using the 2nd version, then I would remove all code needed for the first version.

> Personally I don't think it matters because current implementation is not usable in production and I really doubt anybody is using it. 

@joemcbride said _I certainly hope so since I actively use it._ in https://github.com/graphql-dotnet/graphql-dotnet/pull/3144#issuecomment-1131898717

> Just to clarify, separate project in the repo or completely separate ...

No new repo, no. GraphQL.Federation project here.

_____
So far I re-read all this thread from May and continue to read about federation.

dariuszkuc commented:
Hello 👋 
@Mithras any chance on addressing those PR comments/rebasing to latest in the near term?

Mithras commented:
Hi,
I've changed teams and don't have any Federated services here for now so refactoring this PR wasn't a priority. I'll try to rebase/refactor it this or maybe next week.

Shane32 commented:
If @sungam3r does not have time soon (it sounds like he's got a lot on his plate), then maybe I can review it.  We should try to get this in v8.

sungam3r commented:
Definetly in v8.

dariuszkuc commented:
BTW just an FYI up but I recently externalized the Federation Compatibility Tests -> you can now run them locally using NPX (https://www.npmjs.com/package/@apollo/federation-subgraph-compatibility) or through Github Actions (https://github.com/marketplace/actions/apollo-federation-subgraph-compatibility).

Example GH action integration (workflow is invoked on the PRs) -> https://github.com/apollographql/federation-jvm/blob/master/.github/workflows/federation-compatibility.yaml#L35
Action will then comment on the PRs -> https://github.com/apollographql/federation-jvm/pull/285#issuecomment-1374047449

Mithras commented on src/GraphQL/Federation/Extensions/FederationQuerySchemaNodeVisitor.cs:
```diff
@@ -0,0 +1,69 @@
+            }
+
+            var rep = repMap!.ToObject(resolver.SourceType);
+            var result = resolver.Resolve(context, rep);
```
doesn't work with `ValueTask<T>`. See reply above.

Mithras commented:
@sungam3r 

> How does it feel on NET7 ?

Feels pretty good actually. Omnisharp works out of the box which has never happened with previous .NET versions.

>  I will prefer to accept backward incompatible changes than to have two different implementations on support.

As I have said, I've added new implementation for Feraration 2 with DataLoader, CodeFirst and SchemaFirst support. I intentionally haven't touched old implementation as there is pretty much zero overlap. Old implementation can be removed in like 10 minutes.

> Regarding VS Code configuration files. Yes, I don't use VS Code and I'm fine to change these files in any way you see fit. Or simply remove them from this repo and add into gitignore.

I let `VS Code` update `tasks.json` and made sure it works.

> @joemcbride said I certainly hope so since I actively use it. in #3144 (comment)

Well I'm not sure how it's possible to use existing implementation which doesn't work with DataLoaders for anything more than Hello World.

---

Anyway, I moved everything to `GraphQL.Federation` and `GraphQL.Federation.Tests` accordingly. There are minimal changes outside of the two.

Mithras commented:
@sungam3r 
> Regarding unit tests with spaghetti string literals. I recommend you to switch to raw string literals that are available since C#11.

done.

sungam3r commented:
@Mithras thanks, I remember about your work and this one in my todo list just after OpenTelemetry PR.

Shane32 commented:
If @sungam3r doesn't have time to merge it for v8, I’ll try to do so. However, v8 isn’t planned to be released anytime soon.

michael-watson commented:
Is there anything I can do to help move this PR along? I'm happy to spend some cycles completing out any last items ❤️

Mithras commented:
There is an integration test that tests most of everything. Anyway, I'm on vacation and probably won't even have internet for 1.5 weeks

Shane32 commented:
No problem @Mithras 

The purpose is to have end to end tests plus have the code double as a sample project. So the tests demonstrate two fully functional GraphQL subgraphs created in GraphQL.NET, and the tests start hosting them at two endpoints and have Apollo Federation create a super graph by querying those endpoints, and then start Apollo Router, followed by actually running a query that pulls data from both subgraphs and verifying the result. I’d like to do the same for code-first federation. Sorry I’ve not looked this PR recently to see what tests exist.

michael-watson commented:
@Shane32 I've been working on bringing all of the Federation code together in one place and I put everything in PR #3723. I can't run the github actions locally because of `NUGET_AUTH_TOKEN`. How can I work around this to properly test my code with what you have today?

Shane32 commented:
> @Shane32 I've been working on bringing all of the Federation code together in one place and I put everything in PR #3723. I can't run the github actions locally because of `NUGET_AUTH_TOKEN`. How can I work around this to properly test my code with what you have today?

The federation tests doesn't use the NUGET_AUTH_TOKEN -- it only uses the GITHUB_TOKEN and only to pull the repository.

See https://github.com/graphql-dotnet/graphql-dotnet/blob/master/.github/workflows/test-code.yml and scroll down to the `testfederation` part.  Basically it just:
1. Builds and spins up a couple of the sample projects
2. Runs `rover supergraph compose`
3. Starts router
4. Runs a few test queries against router
5. Validates the responses

All the rest of the tests you can run from Visual Studio.

Shane32 commented:
I'd suggest just writing a powershell script to perform the same actions, or something like that.  Here's ChatGPT's rough draft:

```powershell
# Change to src directory
Set-Location -Path .\src

# Install dependencies for Sample1
dotnet restore .\GraphQL.Federation.SchemaFirst.Sample1

# Build Sample1 and Sample2
dotnet build --no-restore .\GraphQL.Federation.SchemaFirst.Sample1
dotnet build .\GraphQL.Federation.SchemaFirst.Sample2

# Start Sample1 and Sample2
Start-Process dotnet -ArgumentList "run --no-build --project GraphQL.Federation.SchemaFirst.Sample1"
Start-Process dotnet -ArgumentList "run --no-build --project GraphQL.Federation.SchemaFirst.Sample2"

# Install Apollo Rover CLI
Invoke-RestMethod -Uri 'https://rover.apollo.dev/nix/latest' | Out-File -FilePath '.\rover_install.sh'
Start-Process "sh" -ArgumentList ".\rover_install.sh"

# Install jq
Start-Process "sudo" -ArgumentList "apt-get install -y jq"

# Wait for Sample1 and Sample2 to spin up
Function WaitForService ($url) {
    For ($i=1; $i -le 60; $i++) {
        Write-Host "Request $i to the URL..."
        $response = Invoke-RestMethod -Uri $url -Method Get -StatusCodeVariable statusCode
        if ($statusCode -eq 200) {
            Write-Host "Received 200 response, step completed."
            return
        }
        Write-Host "Did not receive a 200 response, sleeping for 0.5 second..."
        Start-Sleep -Milliseconds 500
    }
    Write-Host "Timed out after 30 seconds, step failed."
    Exit 1
}
WaitForService 'http://localhost:5601'
WaitForService 'http://localhost:5602'

# Change to src/Federation directory
Set-Location -Path .\Federation

# Build supergraph
& "$HOME/.rover/bin/rover" supergraph compose --config .\federation-$env:FEDERATION_VERSION-supergraph-config.yaml --elv2-license=accept | Out-File -FilePath '.\supergraph.graphql'

# Print supergraph
Get-Content -Path .\supergraph.graphql

# Download and Start router
Invoke-RestMethod -Uri 'https://router.apollo.dev/download/nix/latest' | Out-File -FilePath '.\router_install.sh'
Start-Process "sh" -ArgumentList ".\router_install.sh"
Start-Process .\router -ArgumentList "--dev --supergraph .\supergraph.graphql"

# Wait for router to spin up
WaitForService 'http://localhost:8088/health'

# Run GraphQL queries
Invoke-RestMethod -Uri 'http://127.0.0.1:4000' -Method Post -Headers @{"Content-Type" = "application/json"} -InFile '.\federation-request-1.json' | Out-File -FilePath '.\response-1.json'
Invoke-RestMethod -Uri 'http://127.0.0.1:4000' -Method Post -Headers @{"Content-Type" = "application/json"} -InFile '.\federation-request-2.json' | Out-File -FilePath '.\response-2.json'
Invoke-RestMethod -Uri 'http://127.0.0.1:4000' -Method Post -Headers @{"Content-Type" = "application/json"} -InFile '.\federation-request-3.json' | Out-File -FilePath '.\response-3.json'

# Function to compare JSON files
Function CompareJsonFiles ($actualFile, $expectedFile) {
    $actualContent = Get-Content -Path $actualFile | ConvertFrom-Json | ConvertTo-Json
    $expectedContent = Get-Content -Path $expectedFile | ConvertFrom-Json | ConvertTo-Json
    if ($actualContent -eq $expectedContent) {
        Write-Host "$actualFile and $expectedFile are identical."
    } else {
        Write-Host "$actualFile and $expectedFile differ."
        Exit 1
    }
}

# Compare query result 1 to expected response
$actualResponse1 = "actual-response-1.json"
$expectedResponse1 = "expected-response-1.json"
Get-Content -Path "response-1.json" | Out-File -FilePath $actualResponse1
Get-Content -Path "federation-response-1.json" | Out-File -FilePath $expectedResponse1
CompareJsonFiles $actualResponse1 $expectedResponse1

# Compare query result 2 to expected response
$actualResponse2 = "actual-response-2.json"
$expectedResponse2 = "expected-response-2.json"
Get-Content -Path "response-2.json" | Out-File -FilePath $actualResponse2
Get-Content -Path "federation-response-2.json" | Out-File -FilePath $expectedResponse2
CompareJsonFiles $actualResponse2 $expectedResponse2

# Compare query result 3 to expected response
$actualResponse3 = "actual-response-3.json"
$expectedResponse3 = "expected-response-3.json"
Get-Content -Path "response-3.json" | Out-File -FilePath $actualResponse3
Get-Content -Path "federation-response-3.json" | Out-File -FilePath $expectedResponse3
CompareJsonFiles $actualResponse3 $expectedResponse3
```

Shane32 commented:
If you devise a good powershell script, add some comments that explain how to use it, post a PR with the file, and I'll merge it in.

Shane32 commented:
@michael-watson By the way, I'm okay with having a separate NuGet package for federation, or having it in the main GraphQL NuGet package.

Shane32 commented:
Closing in favor of #3921, which extended this PR with additional functionality.  Thank you for your contributions @Mithras !
